message_list_view: Improve highlighting and re-rendering of mentions on name change.

Varunvaruns9 opened the pull request:
#11026 

Add green highlighting to user group mentions.
Improve re-rendering of mentions on user name change and add similar logic for user group/ stream name change.

timabbott commented on static/js/message_list_view.js:
```diff
@@ -486,6 +479,35 @@ MessageListView.prototype = {
+                // inside the highlight.
+                $(this).text("@" + people.get_person_from_user_id(user_id).full_name);
+            }
+        });
```
Let's do this piece as its own commit; also, I think it might be more correc to just remove the `row.hasClass('mention')` if statement, rather than moving this block out.  And then we can not duplicate the `user-group-mention` block too...

Varunvaruns9 commented:
@timabbott Changed the commits.

timabbott commented:
It looks like your first commits adds a `.length` in the highlight section; can you explain that?  

Please don't make extra changes, even if they're fixes for bugs, squashed into commits that move code around; it makes life very difficult for reviewers.  You should do a separate commit for the `.length` change if that's desired.

timabbott commented:
Otherwise, this lgtm.

timabbott commented:
(It'd be nice to have `node` unit tests for this functionality; I considered adding them in the original work , but it's kinda difficult to do the setup properly, since we need to render full message templates with full message content, and there's a lot of stubbing one ends up needing).

Varunvaruns9 commented:
Yes, sorry I added `.length` as the condition `!$(this).find(".highlight")` came out to be `false` every time as `.find()` returns an object; we need to check if `length` property has value 0 to see if there was any matching element.
I made the change and forgot to add it as a separate commit. I will change it and add the necessary commit.

Also I replaced class `mention` with `message_row` instead of removing it completely as removing the if condition causes some casper tests to fail and since `message_row` is present in every row, it seemed more correct.

Varunvaruns9 commented:
@timabbott ping.

timabbott commented:
I merged the first commit, thanks for extracting it.  Note that I'm in SF timezone, so I was asleep for your earlier comment; generally there's no need to ping me about something unless it's been a few days since your last comment.

Can you give details on the casper failures with `message_row`?  I would have thought everything passed into that function was a  `message_row`...

timabbott commented:
(Otherwise, the rest of this PR lgtm)

Varunvaruns9 commented:
When I removed the if condition, the casper tests were showing some transient behaviour, sometimes failing at `00-realm-creation` and sometimes at `10-admin`. Also, CircleCI tests failed last time on my commits at `10-admin`. I tried it just now without the if, here is what I got:
```
FAIL "form[action^="/accounts/register/"]" never disappeared in 5000ms
    type: uncaughtError
    file: /srv/zulip/frontend_tests/casper_tests/00-realm-creation.js
    error: "form[action^="/accounts/register/"]" never disappeared in 5000ms

    stack: not provided
PASS Untitled suite in /srv/zulip/frontend_tests/casper_tests/00-realm-creation.js
FAIL 5 tests executed in 6.874s, 4 passed, 1 failed, 0 dubious, 0 skipped.      

Details for the 1 failed test:

In /srv/zulip/frontend_tests/casper_tests/00-realm-creation.js
  Untitled suite in /srv/zulip/frontend_tests/casper_tests/00-realm-creation.js
    uncaughtError: "form[action^="/accounts/register/"]" never disappeared in 5000ms
```

timabbott commented:
I tested locally, and did not see casper test failures with replacing the `if` with `if true`.  Pushed a test commit here that removes the if statement entirely.

Varunvaruns9 commented:
All tests passed, maybe something got fixed since the last time they ran, although the casper tests still fail occasionally on my system without any changes.

timabbott commented:
Merged as the series of commits ending with 5d3cf2e40a3ca27ab4547b20adce503f3d81b860, after a bit of rearranging to be maximally readable.  Thanks @Varunvaruns9!

timabbott commented:
If you have a really slow system, that often exposes nondeterministic failures in the Casper test suite.  The issues are likely present in master, and worth debugging (though Casper tests can be pretty tricky to debug).
